nft/collection/sign_process.go

```diff
@@ -0,0 +1,318 @@
+package collection
+
+import (
+	"sync"
+
+	extensioncurrency "github.com/ProtoconNet/mitum-currency-extension/currency"
+	"github.com/ProtoconNet/mitum-nft/nft"
+	"github.com/pkg/errors"
+	"github.com/spikeekips/mitum-currency/currency"
+	"github.com/spikeekips/mitum/base"
+	"github.com/spikeekips/mitum/base/operation"
+	"github.com/spikeekips/mitum/base/state"
+	"github.com/spikeekips/mitum/util/valuehash"
+)
+
+var SignItemProcessorPool = sync.Pool{
+	New: func() interface{} {
+		return new(SignItemProcessor)
+	},
+}
+
+var SignProcessorPool = sync.Pool{
+	New: func() interface{} {
+		return new(SignProcessor)
+	},
+}
+
+func (Sign) Process(
+	func(key string) (state.State, bool, error),
+	func(valuehash.Hash, ...state.State) error,
+) error {
+	return nil
+}
+
+type SignItemProcessor struct {
+	cp     *extensioncurrency.CurrencyPool
+	h      valuehash.Hash
+	nft    nft.NFT
+	nst    state.State
+	sender base.Address
+	item   SignItem
+}
+
+func (ipp *SignItemProcessor) PreProcess(
+	getState func(key string) (state.State, bool, error),
+	_ func(valuehash.Hash, ...state.State) error,
+) error {
+
+	if err := ipp.item.IsValid(nil); err != nil {
+		return err
+	}
+
+	nid := ipp.item.NFT()
+
+	// check collection
+	if st, err := existsState(StateKeyCollection(nid.Collection()), "design", getState); err != nil {
+		return err
+	} else if design, err := StateCollectionValue(st); err != nil {
+		return err
+	} else if !design.Active() {
+		return errors.Errorf("dead collection; %q", nid.Collection())
+	}
+
+	var holders []nft.Signer
+	var n nft.NFT
+
+	// check nft
+	if st, err := existsState(StateKeyNFT(nid), "nft", getState); err != nil {
+		return err
+	} else if nv, err := StateNFTValue(st); err != nil {
+		return err
+	} else {
+		switch ipp.item.Qualification() {
+		case CreatorQualification:
+			holders = nv.Creators()
+		case CopyrighterQualification:
+			holders = nv.Copyrighters()
+		default:
+			return errors.Errorf("wrong qualification; %q", ipp.item.Qualification())
+		}
+		n = nv
+		ipp.nst = st
+	}
+
+	// check owner
+	if n.Owner().String() == "" {
+		return errors.Errorf("dead nft; %q", nid)
+	}
+
+	var idx = -1
+	for i := range holders {
+		if holders[i].Account().Equal(ipp.sender) {
+			idx = i
```

Comment:
@wyuinche 
Why is the use of idx necessary? I think it is enough to just return an error right inside the for statement.